fix(epics): use actual group_id for save/delete operations on nested epics

JohnVillalovos opened the pull request:
When an epic belonging to a subgroup is retrieved through a parent
group's epic listing, save() and delete() operations would fail because
they used the parent group's path instead of the epic's actual group_id.

Add an internal custom path hook for object save and delete operations, and let
UpdateMixin and DeleteMixin accept an explicit `_custom_path` keyword. GroupEpic
uses this hook to build mutation paths from the epic's server-provided group_id,
so save() and delete() target the epic's owning group.

Avoid using parent manager attrs as a fallback for the epic group path. Lazy
epics do not have a server-provided group_id, so raise a clear error instead of
silently using the parent group path.

Add unit coverage for custom update/delete paths, GroupEpic save/delete path
selection, and lazy epic path failures. Add a functional test for saving a
subgroup epic discovered through the parent group listing.

Assisted-by: OpenAI Codex (GPT-5)

Closes: #3261

nejch commented on gitlab/v4/objects/epics.py:
```diff
@@ -29,6 +30,63 @@ class GroupEpic(ObjectDeleteMixin, SaveMixin, RESTObject):
 
+    def _epic_path(self) -> str:
+        """Return the API path for this epic using its real group."""
+        if not hasattr(self, "group_id") or self.group_id is None:
```
Can a `group_id` actually be none for an epic, since it's always a group resource? Or can we go with a simple try/except here?

One other thought - will this work if we get an epic with `lazy=True` from an arbitrary group?

JohnVillalovos replied:
@nejch Good catch. `lazy=True` would fail.

Will work on an update.

JohnVillalovos replied:
Fixed.

nejch commented on gitlab/v4/objects/epics.py:
```diff
@@ -29,6 +30,63 @@ class GroupEpic(ObjectDeleteMixin, SaveMixin, RESTObject):
+        path = self._epic_path()
+
+        # Call SaveMixin.save() method
+        return super().save(_pg_custom_path=path, **kwargs)
```
Not sure if we should have some more generic `_get_custom_path` or `_resolve_path` helpers that would automatically get picked up by the mixins, as we already do this twice for delete/update. That way we don't need an override here (or maybe also in some other places), but I'm not sure if feasible or worth it.

JohnVillalovos replied:
I added `_get_custom_path()` method.

nejch commented on tests/functional/api/test_epics.py:
```diff
@@ -32,3 +37,42 @@ def test_epic_notes(epic):
+        }
+    )
+
+    nested_epic = subgroup.epics.create(
```
Could we have this one as a fixture that `yield`s the epic, to avoid the manual cleanup below?

Or I'm just confused by the try/finally block, maybe it was meant to test deletion even if previous things failed.

JohnVillalovos replied:
Done.

nejch reviewed (commented):
Thanks @JohnVillalovos, looks good and that's a tricky one. Not sure about lazily fetched epics - just have a few questions there.

JohnVillalovos commented on gitlab/mixins.py:
```diff
@@ -314,6 +314,8 @@ def update(
             path = self.path
         else:
             path = f"{self.path}/{utils.EncodedId(id)}"
+        if "_pg_custom_path" in kwargs:
```
Changed the code so no longer have this.

JohnVillalovos committed:
fix(epics): use actual group_id for save/delete operations on nested epics

When an epic belonging to a subgroup is retrieved through a parent
group's epic listing, save() and delete() operations would fail because
they used the parent group's path instead of the epic's actual group_id.

Add an internal custom path hook for object save and delete operations, and let
UpdateMixin and DeleteMixin accept an explicit `_custom_path` keyword. GroupEpic
uses this hook to build mutation paths from the epic's server-provided group_id,
so save() and delete() target the epic's owning group.

Avoid using parent manager attrs as a fallback for the epic group path. Lazy
epics do not have a server-provided group_id, so raise a clear error instead of
silently using the parent group path.

Add unit coverage for custom update/delete paths, GroupEpic save/delete path
selection, and lazy epic path failures. Add a functional test for saving a
subgroup epic discovered through the parent group listing.

Assisted-by: OpenAI Codex (GPT-5)

Closes: #3261
